### PR TITLE
fix(client): Correct verbosity of correlation warnings and fix the callback context map

### PR DESCRIPTION
**Pull request by jamdavi**
This addresses issue #1349 and fixes the following problems

* Too many warning messages for the correlation callback not being set, which is optional
* Fix NPE if no callback context is specified
* Fix NPE if correlationId is null

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 1]
```diff
@@ -242,7 +242,7 @@ public void onMessageSent(Message message, String deviceId, Throwable e)
             try
             {
                 String messageId = message.getCorrelationId();
-                if (messageId != null && correlationCallbacks.containsKey(messageId))
```
Comment (drwill-ms) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:245]: Why remove `messageId != null`? If null an exception would be thrown.

Comment (drwill-ms) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:245]: https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2.containskey?view=net-5.0#exceptions

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 2]
```diff
@@ -242,7 +242,7 @@ public void onMessageSent(Message message, String deviceId, Throwable e)
             try
             {
                 String messageId = message.getCorrelationId();
```
Comment (drwill-ms) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:244]: Our variable name is `messageId`, but the method is to get a correlation Id. Should we rename the variable to match?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java (hunk 1 above)
Comment (jamdavi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:245]: Actually yes, you're right. I'm using a concurrent hashmap so supplying a null will throw an NPE.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java (hunk 2 above)
Comment (jamdavi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:244]: I'm good with this.

**Review by timtay-microsoft — APPROVED** [review]

**Review by drwill-ms — APPROVED** [review]

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java [hunk 3]
```diff
@@ -1704,12 +1700,15 @@ private void addToWaitingQueue(IotHubTransportPacket packet)
...
-                String messageId = message.getCorrelationId();
-                if (!correlationCallbacks.containsKey(messageId))
+                String correlationId = message.getCorrelationId();
+                if (correlationId != null && !correlationCallbacks.containsKey(correlationId))
                 {
-                    correlationCallbacks.put(messageId, message.getCorrelatingMessageCallback());
-                    correlationCallbackContexts.put(messageId, message.getCorrelatingMessageCallbackContext());
-                    correlationCallbacks.get(messageId).onRequestQueued(message, packet, correlationCallbackContexts.get(messageId));
+                    correlationCallbacks.put(correlationId, message.getCorrelatingMessageCallback());
+                    if (message.getCorrelatingMessageCallbackContext() != null)
```
Comment (drwill-ms) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:1707]: Consider calling and caching the result, so you don't have to call it twice.

Comment (DBY047) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:1729]: I don't see the point of adding this check. message.getCorrelationId(); can't return null. 
```java
    /**
     * Getter for the correlationId property
     * @return The property value
     */
    public String getCorrelationId()    
    {
        // Codes_SRS_MESSAGE_34_045: [The function shall return the message's correlation ID.]
        if (correlationId == null)
        {
            return "";
        }

        return correlationId;
    }
```

Comment (jamdavi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:1729]: Can you link to that? I am looking in our current master and that's not there.

Comment (jamdavi) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java:1729]: Ah I see, yes, that method is changed on my end (as part of a larger refactor) to a Lombok Getter/Setter but it isn't guarded with @NonNull. Thanks for catching that as I wasn't in the master branch when I was looking at the code.

I've updated the branch I'm working on to keep this behavior.

**Pull request merged**